src/distdl/nn/conv_feature_new.py

```diff
@@ -0,0 +1,388 @@
+        self.torch_halo_pad = to_torch_pad(halo_shape)
+
+        # Determine the halo shapes of the neighboring ranks
+        neighbor_halo_shapes = self.P_x.broadcast_to_neighbors(halo_shape)
+
+        # We can now compute the info required for the halo layer.
+        recv_buffer_shape, send_buffer_shape = self._compute_exchange_info(halo_shape, neighbor_halo_shapes)
+
+        # We can also set up part of the halo layer.
+        self.halo_layer = HaloExchange(self.P_x,
+                                       halo_shape,
+                                       recv_buffer_shape,
+                                       send_buffer_shape,
+                                       buffer_manager=self.buffer_manager)
+
+        # We have to select out the "unused" entries.  Sometimes there can
+        # be "negative" halos.
+        needed_ranges = self._compute_needed_ranges(x_local_shape_after_pad, halo_shape_with_negative)
+        self.needed_slices = assemble_slices(needed_ranges[:, 0],
+                                             needed_ranges[:, 1])
+
+    def _distdl_module_teardown(self, input):
+        r"""Distributed (channel) convolution module teardown function.
+
+        This function is called every time something changes in the input
+        tensor structure.  It should not be called manually.
+
+        Parameters
+        ----------
+        input :
+            Tuple of forward inputs.  See
+            `torch.nn.Module.register_forward_pre_hook` for more details.
+
+        """
+
+        # Reset all sub_layers
+        self.pad_layer = None
+        self.torch_halo_pad = None
+        self.needed_slices = None
+        self.halo_layer = None
+
+        # Reset any info about the input
+        self._distdl_is_setup = False
+        self._input_tensor_structure = TensorStructure()
+
+    def _distdl_input_changed(self, input):
+        r"""Determine if the structure of inputs has changed.
+
+        Parameters
+        ----------
+        input :
+            Tuple of forward inputs.  See
+            `torch.nn.Module.register_forward_pre_hook` for more details.
+
+        """
+
+        new_tensor_structure = TensorStructure(input[0])
+
+        return self._input_tensor_structure != new_tensor_structure
+
+    def forward(self, input):
+        r"""Forward function interface.
+
+        Parameters
+        ----------
+        input :
+            Input tensor to be broadcast.
+
+        """
+
+        if not self.P_x.active:
+            return input.clone()
+
+        if self.serial:
+            return self.conv_layer(input)
+
+        w = self.w_broadcast(self.weight)
+        self.conv_layer.weight = w
+
+        if self.conv_layer.bias is not None:
+            b = self.b_broadcast(self.bias)
+            self.conv_layer.bias = b
+
+        input_padded = self.pad_layer(input)
+        input_padded_with_halo = F.pad(input_padded, pad=self.torch_halo_pad, mode='constant', value=0)
```

Comment:
If you are calling `F.pad()` directly, why is there `input_padded`?

Reply:
See below, but if I understand correctly, this pattern applies only to boundary workers *if* there is torch padding.  That should be documented here (ideally self documenting by only calling the pad layer on those workers).

Reply:
Having two padding layers is indeed wasteful. I will merge the two.